training_record/services.py

```diff
@@ -79,6 +79,76 @@ def create_off_campus_record_from_raw_data(
                 )
         return record
 
+    # pylint: disable=invalid-name
+    # pylint: disable=redefined-builtin
+    # pylint: disable=unused-argument
+    @staticmethod
+    def update_off_campus_record_from_raw_data(
+            record, off_campus_event=None, user=None,
+            contents=None, attachments=None):
+        '''Update the off-campus record
+
+        Parameters
+        ----------
+        off_campus_event: dict
+            This dict should have full information needed to update an
+            OffCampusEvent.
+        user: User
+            The user of which the record is related to.
+        contents(optional): list of dict
+            Every dict of this list should have full information needed to
+            create a RecordContent.
+        attachments(optional): list of InMemoryFile
+
+        Returns
+        -------
+        record: Record
+        '''
+        off_campus_event_data = off_campus_event
+        if off_campus_event_data is None:
+            raise BadRequest('校外培训活动数据格式无效')
+        if contents is None:
+            contents = []
+        if attachments is None:
+            attachments = []
+
+        with transaction.atomic():
+            # get the record to be updated
+            try:
+                record = Record.objects.get(id=record.id)
+            except Exception:
+                raise BadRequest('校外培训记录无效')
+
+            # update the offCampusEvent
+            try:
+                off_campus_event_instance = (
+                    OffCampusEvent.objects.select_for_update().get(
+                        id=off_campus_event_data.get('id'))
+                )
+                for key, val in off_campus_event_data.items():
+                    setattr(off_campus_event_instance, key, val)
+                off_campus_event_instance.save()
+
+            except Exception:
+                raise BadRequest('校外培训活动数据格式无效')
+
+            # add attachments
+            for attachment in attachments:
```

Comment:
After the creation of attachments, check the number of related attachments, and raise exception if there are too many attachments.